daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -1503,115 +1570,129 @@ private[lf] final class Compiler(
+                  ),
+                ) in SEVar(2) // mb contract id
+              }
+            )
           }
-        )
-      }
-    }
-  }
+        ))
 
   private[this] val fetchByKeyResultStruct: Struct[Int] =
     Struct.assertFromSeq(List(contractIdFieldName, contractFieldName).zipWithIndex)
 
   @inline
-  private[this] def compileFetchByKey(templateId: TypeConName, key: SExpr) = {
+  private[this] def compileFetchByKey(
+      tmplId: TypeConName,
+      tmpl: Template,
+      tmplKey: TemplateKey,
+  ): (FetchByKeyDefRef, SExpr) =
     // Translates 'fetchByKey Foo <key>' into:
     // let keyWithMaintainers = {key: <key>, maintainers: <key maintainers> <key>}
     // in \token ->
     //    let coid = $fetchKey keyWithMaintainers token
     //        contract = $fetch coid token
     //        _ = $insertFetch coid <signatories> <observers> Some(keyWithMaintainers)
     //    in { contractId: ContractId Foo, contract: Foo }
-    val template = lookupTemplate(templateId)
-    withEnv { _ =>
-      val keyTemplate = template.key.getOrElse(
-        throw CompilationError(
-          s"Expecting to find key for template ${templateId}, but couldn't",
-        ),
-      )
-      SELet(encodeKeyWithMaintainers(key, keyTemplate)) in {
-        env = env.incrPos // key with maintainers
-        withLabel(
-          s"fetchByKey @${templateId.qualifiedName}",
-          SEAbs(1) {
-            env = env.incrPos // token
-            env = env.addExprVar(template.param)
-            // TODO should we evaluate this before we even construct
-            // the update expression? this might be better for the user
-            val signatories = compile(template.signatories)
-            val observers = compile(template.observers)
-            SELet(
-              SBUFetchKey(templateId)(
-                SEVar(2), // key with maintainers
-                SEVar(1) // token
-              ),
-              SBUFetch(templateId)(
-                SEVar(1), /* coid */
-                SEVar(2) /* token */
-              ),
-              SBUInsertFetchNode(templateId, byKey = true)(
-                SEVar(2), // coid
-                signatories,
-                observers,
-                SEApp(SEBuiltin(SBSome), Array(SEVar(4))),
-                SEVar(3) // token
-              ),
-            ) in SBStructCon(fetchByKeyResultStruct)(
-              SEVar(3), // contract id
-              SEVar(2) // contract
+    FetchByKeyDefRef(tmplId) ->
+      validate(
+        closureConvert(
+          Map.empty,
+          withEnv { _ =>
+            withLabel(
+              s"fetchByKey @${tmplId.qualifiedName}",
+              SEAbs(2) {
+                env = env.incrPos // key
+                env = env.incrPos // token
+                env = env.addExprVar(tmpl.param)
+                // TODO should we evaluate this before we even construct
+                // the update expression? this might be better for the user
```

Comment:
Does this comment make any sense? There's nothing we could evaluate before we fetched the contract from the ledger, which needs to happen in the `Update` monad, because we need the payload of the contract for all the evaluations. Or am I misunderstanding something?

Reply:
dropped.

Reply:
Merci.